chisel/src/cli.rs

```diff
@@ -0,0 +1,200 @@
+//! Chisel CLI
+//!
+//! This module contains the core readline loop for the Chisel CLI as well as the
+//! executable's `main` function.
+
+use chisel::{
+    prelude::{ChiselCommand, ChiselDisptacher, DispatchResult},
+    solidity_helper::SolidityHelper,
+};
+use clap::Parser;
+use foundry_cli::cmd::{forge::build::BuildArgs, LoadConfig};
+use foundry_common::evm::EvmArgs;
+use foundry_config::{
+    figment::{
+        value::{Dict, Map},
+        Metadata, Profile, Provider,
+    },
+    Config,
+};
+use rustyline::{error::ReadlineError, Editor};
+use yansi::Paint;
+
+// Loads project's figment and merges the build cli arguments into it
+foundry_config::merge_impl_figment_convert!(ChiselParser, opts, evm_opts);
+
+/// Version tag generated by vergen
+pub(crate) const VERSION_MESSAGE: &str = concat!(
+    env!("CARGO_PKG_VERSION"),
+    " (",
+    env!("VERGEN_GIT_SHA_SHORT"),
+    " ",
+    env!("VERGEN_BUILD_TIMESTAMP"),
+    ")"
+);
+
+/// Chisel is a fast, utilitarian, and verbose solidity REPL.
+#[derive(Debug, Parser)]
+#[clap(name = "chisel", version = VERSION_MESSAGE)]
+pub struct ChiselParser {
+    #[clap(flatten, next_help_heading = "BUILD OPTIONS")]
+    pub opts: BuildArgs,
+
+    #[clap(flatten, next_help_heading = "EVM OPTIONS")]
+    pub evm_opts: EvmArgs,
+
+    #[command(subcommand)]
+    pub sub: Option<ChiselParserSub>,
+}
+
+/// Chisel binary subcommands
+#[derive(clap::Subcommand, Debug)]
+pub enum ChiselParserSub {
+    /// List all cached sessions
+    List,
+    /// Load a cached session
+    Load { id: String },
+    /// View the source of a cached session
+    View { id: String },
+    /// Clear all cached chisel sessions from the cache directory
+    ClearCache,
+}
+
+#[tokio::main]
+async fn main() {
+    // Parse command args
+    let args = ChiselParser::parse();
+
+    // Keeps track of whether or not an interrupt was the last input
+    let mut interrupt = false;
+
+    // Create a new rustyline Editor
+    let mut rl = Editor::<SolidityHelper>::new().unwrap_or_else(|e| {
+        tracing::error!(target: "chisel-env", "Failed to initialize rustyline Editor! {}", e);
+        panic!("Failed to create a rustyline Editor for the chisel environment! {e}");
+    });
+    rl.set_helper(Some(SolidityHelper));
+
+    let (config, evm_opts) = args.load_config_and_evm_opts().unwrap_or_else(|e| {
+        tracing::error!(target: "chisel-env", "Failed to load Foundry config! {}", e);
+        panic!("Failed to load foundry config! {}", e)
+    });
+
+    // Create a new cli dispatcher
+    let mut dispatcher = ChiselDisptacher::new(&chisel::session_source::SessionSourceConfig {
+        // Enable traces if any level of verbosity was passed
+        traces: config.verbosity > 0,
+        foundry_config: config,
+        evm_opts,
+        backend: None,
+    });
+
+    // Print intro header
+    println!("Welcome to Chisel! Type `{}` to show available commands.", Paint::green("!help"));
```

Comment:
This needs to go below the subcommand match
![Screenshot from 2022-11-26 13-57-14](https://user-images.githubusercontent.com/8406232/204104818-51ae7da1-211f-468b-a148-fe314f1a562f.png)

Reply:
resolved